src/services/ghost/GhostStreamingParser.ts

```diff
@@ -0,0 +1,510 @@
+
+		return {
+			suggestions,
+			isComplete,
+			hasNewSuggestions,
+		}
+	}
+
+	/**
+	 * Mark the stream as finished and process any remaining content with sanitization
+	 */
+	public finishStream(): StreamingParseResult {
+		this.streamFinished = true
+		return this.processChunk("")
+	}
+
+	/**
+	 * Extract completed <change> blocks from the buffer
+	 */
+	private extractCompletedChanges(): ParsedChange[] {
+		const newChanges: ParsedChange[] = []
+
+		// Look for complete <change> blocks starting from where we left off
+		const searchText = this.buffer.substring(this.lastProcessedIndex)
+
+		// Updated regex to handle both single-line XML format and traditional format with whitespace
+		const changeRegex =
+			/<change>\s*<search>\s*<!\[CDATA\[([\s\S]*?)\]\]>\s*<\/search>\s*<replace>\s*<!\[CDATA\[([\s\S]*?)\]\]>\s*<\/replace>\s*<\/change>/g
+
+		let match
+		let lastMatchEnd = 0
+
+		while ((match = changeRegex.exec(searchText)) !== null) {
+			// Preserve cursor marker in search content (LLM includes it when it sees it in document)
+			const searchContent = match[1]
+			// Extract cursor position from replace content
+			const replaceContent = match[2]
+			const cursorPosition = this.extractCursorPosition(replaceContent)
+
+			newChanges.push({
+				search: searchContent,
+				replace: replaceContent,
+				cursorPosition,
+			})
+
+			lastMatchEnd = match.index + match[0].length
+		}
+
+		// Update our processed index to avoid re-processing the same content
+		if (lastMatchEnd > 0) {
+			this.lastProcessedIndex += lastMatchEnd
+		}
+
+		return newChanges
+	}
+
+	/**
+	 * Check if the response appears to be complete
+	 */
+	private isResponseComplete(): boolean {
+		// Simple heuristic: if we haven't seen new content for a while and
+		// the buffer doesn't end with an incomplete tag, consider it complete
+		const trimmedBuffer = this.buffer.trim()
+
+		// Check if we have any incomplete <change> tags
+		const incompleteChangeMatch = /<change(?:\s[^>]*)?>(?:(?!<\/change>)[\s\S])*$/i.test(trimmedBuffer)
+		const incompleteSearchMatch = /<search(?:\s[^>]*)?>(?:(?!<\/search>)[\s\S])*$/i.test(trimmedBuffer)
+		const incompleteReplaceMatch = /<replace(?:\s[^>]*)?>(?:(?!<\/replace>)[\s\S])*$/i.test(trimmedBuffer)
+		const incompleteCDataMatch = /<!\[CDATA\[(?:(?!\]\]>)[\s\S])*$/i.test(trimmedBuffer)
+
+		// If we have incomplete tags, the response is not complete
+		if (incompleteChangeMatch || incompleteSearchMatch || incompleteReplaceMatch || incompleteCDataMatch) {
+			return false
+		}
+
+		// If the buffer is empty or only whitespace, consider it complete
+		if (trimmedBuffer.length === 0) {
+			return true
+		}
+
+		// If we have at least one complete change and no incomplete tags, likely complete
+		return this.completedChanges.length > 0
+	}
+
+	/**
+	 * Generate suggestions from completed changes
+	 */
+	private generateSuggestions(changes: ParsedChange[]): GhostSuggestionsState {
```

Comment:
_should_ the parser create the suggestions? I feel like this function should live somewhere else?